File: src/time/clock.ts
/**
 * Civil-time helpers. Every date-sensitive calculation resolves in the user's
 * own timezone (§5). A due date becomes overdue at local midnight, never at
 * 00:00 UTC — see fixture T12.
 */

/** An instant in time, epoch milliseconds. */
export type Instant = number;

/** A civil date with no time or zone attached, rendered as `YYYY-MM-DD`. */
export type LocalDate = string & { readonly __localDate: unique symbol };

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;

export function localDate(value: string): LocalDate {
  if (!ISO_DATE.test(value)) throw new Error(`Malformed local date: ${value}`);
  return value as LocalDate;
}

const formatters = new Map<string, Intl.DateTimeFormat>();

function formatterFor(timeZone: string): Intl.DateTimeFormat {
  let f = formatters.get(timeZone);
  if (!f) {
    f = new Intl.DateTimeFormat('en-US', {
      timeZone,
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    });
    formatters.set(timeZone, f);
  }
  return f;
}

/** Resolve an instant to the civil date observed in `timeZone`. */
export function toLocalDate(at: Instant, timeZone: string): LocalDate {
  const parts = formatterFor(timeZone).formatToParts(new Date(at));
  const get = (t: Intl.DateTimeFormatPartTypes): string => {
    const p = parts.find((x) => x.type === t);
    if (!p) throw new Error(`Timezone ${timeZone} produced no ${t} part`);
    return p.value;
  };
  return localDate(`${get('year')}-${get('month')}-${get('day')}`);
}

/** `YYYY-MM-DD` sorts lexicographically, which matches chronological order. */
export function compareDates(a: LocalDate, b: LocalDate): -1 | 0 | 1 {
  return a < b ? -1 : a > b ? 1 : 0;
}

export const isBefore = (a: LocalDate, b: LocalDate): boolean => a < b;
export const isAfter = (a: LocalDate, b: LocalDate): boolean => a > b;
export const isOnOrBefore = (a: LocalDate, b: LocalDate): boolean => a <= b;

/**
 * A claim is overdue once its due date has fully passed in the user's own
 * timezone: due on the 1st means overdue from local midnight starting the 2nd.
 */
export const isOverdue = (dueDate: LocalDate, today: LocalDate): boolean => today > dueDate;

File: app/lib/engine/clock.dart
/// Civil-time helpers. Every date-sensitive calculation resolves in the user's
/// own timezone (§5): a due date turns overdue at local midnight, never at
/// 00:00 UTC.
library;

/// A civil date with no time or zone attached.
class LocalDate implements Comparable<LocalDate> {
  const LocalDate(this.year, this.month, this.day);

  factory LocalDate.parse(String iso) {
    final m = RegExp(r'^(\d{4})-(\d{2})-(\d{2})$').firstMatch(iso);
    if (m == null) throw ArgumentError('Malformed local date: $iso');
    return LocalDate(
      int.parse(m.group(1)!),
      int.parse(m.group(2)!),
      int.parse(m.group(3)!),
    );
  }

  /// Resolve an instant to the civil date observed in [offset].
  ///
  /// Dart has no IANA timezone database in the core SDK, so the zone is
  /// supplied as the offset in effect at that instant. Callers holding a real
  /// IANA zone resolve the offset first; this keeps the engine free of a
  /// timezone dependency while preserving local-midnight semantics.
  factory LocalDate.at(DateTime instant, Duration offset) {
    final shifted = instant.toUtc().add(offset);
    return LocalDate(shifted.year, shifted.month, shifted.day);
  }

  final int year;
  final int month;
  final int day;

  LocalDate addDays(int days) {
    final d = DateTime.utc(year, month, day).add(Duration(days: days));
    return LocalDate(d.year, d.month, d.day);
  }

  int differenceInDays(LocalDate other) =>
      DateTime.utc(year, month, day)
          .difference(DateTime.utc(other.year, other.month, other.day))
          .inDays;

  @override
  int compareTo(LocalDate other) {
    if (year != other.year) return year.compareTo(other.year);
    if (month != other.month) return month.compareTo(other.month);
    return day.compareTo(other.day);
  }

  bool operator <(LocalDate other) => compareTo(other) < 0;
  bool operator >(LocalDate other) => compareTo(other) > 0;
  bool operator <=(LocalDate other) => compareTo(other) <= 0;
  bool operator >=(LocalDate other) => compareTo(other) >= 0;

  @override
  bool operator ==(Object other) =>
      other is LocalDate &&
      other.year == year &&
      other.month == month &&
      other.day == day;

  @override
  int get hashCode => Object.hash(year, month, day);

  @override
  String toString() => '${year.toString().padLeft(4, '0')}'
      '-${month.toString().padLeft(2, '0')}'
      '-${day.toString().padLeft(2, '0')}';
}

/// A claim is overdue once its due date has fully passed locally: due on the
/// 1st means overdue from local midnight starting the 2nd.
bool isOverdue(LocalDate dueDate, LocalDate today) => today > dueDate;
